### ISSUE TITLE
odo create --starter flag does not parse arguments correctly

### ISSUE THREAD

**Issue by valaparthvi** [openshift/odo#5014]
## What versions of software are you using?

**Operating System:**
Fedora 34
**Output of `odo version`:**
odo v2.3.0
## How did you run odo exactly?
1. mkdir /tmp/comp; cd /tmp/comp
2. odo create nodejs mynode --starter=ffff
3. odo create nodejs mynode --starter fffff
4. odo create nodejs --starter ffff

## Actual behavior
1.
2. Fails. Error - `the project: ffff specified in --starter does not exist, available projects: nodejs-starter`
3. Fails. Error - `accepts between 0 and 2 arg(s), received 3`
3. Passes, by taking `fffff` as the component name and downloading the first available starter project.

## Expected behavior
1.
2. Works/Fails as expected, but #4939.
3. Works/Fails as expected, but it should fail with the same error as 2.
4. Should fail, but with the same error as 2 and should not take the value of starter project (i.e. fffff) as the component name.

## Any logs, error output, etc?
```sh
➜ mkcd /tmp/comp
➜  odo create nodejs mynod --starter=ffff     
Devfile Object Validation
 ✓  Checking devfile existence [109850ns]
 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [180315ns]
Validation
 ✓  Validating if devfile name is correct [168525ns]
 ✗  the project: ffff specified in --starter does not exist, available projects: nodejs-starter
➜  rm devfile.yaml
➜  odo create nodejs mynod --starter ffff
Error: accepts between 0 and 2 arg(s), received 3


➜ odo create nodejs --starter ffff
Devfile Object Validation
 ✓  Checking devfile existence [48480ns]
 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [94882ns]
Validation
 ✓  Validating if devfile name is correct [80889ns]

Starter Project
 ✓  Downloading starter project nodejs-starter from https://github.com/odo-devfiles/nodejs-ex.git [721ms]

Please use `odo push` command to create the component with source deployed
```

**Comment by feloy** [comment]
This is a limitation of cobra/pflag.

See https://github.com/spf13/pflag/issues/321, https://github.com/spf13/pflag/issues/134 (https://github.com/spf13/pflag/issues/134#issuecomment-321541601)

tl;dr
On (3) (`odo create nodejs mynode --starter fffff`), there is an ambiguity if fffff is the value for starter, or the third positional parameter, because the value of starter is optional. cobra/pflag historically chose to consider it the positional parameter, and don't want to change, or it would break the current behaviour.

I would recommend to not rely on this behaviour (NoOptDefVal), but find another solution to make a starter optional.

A suggestion:
- `--starter` is a boolean flag
- `--starter-project` is an optional string flag
- (3) becomes `odo create nodejs mynode --starter --starter-project fffff`

**Comment by feloy** [comment]
After discussion with the team:
- `--starter` value is not optional anymore (do not use NoOptDefVal behaviour)
- if a value for `--starter` is not provided, odo can provide the list of possible values (and indicate which one is the *default* one)

**Comment by dharmit** [comment]
If user has not provided starter project name, we error out, along with the list of possible values.

**Comment by mohammedzee1000** [comment]
Started working on this in ernest. Looking at how to handle corbra error for starter flag

**Comment by mohammedzee1000** [comment]
ok so I have one doubt here.

We can handle the flag probem by
1. removing the `NoOptDefVal` and 
2. using the cobra `func (c *Command) SetFlagErrorFunc(f func(*Command, error) error)` function.

So the way this function works is it gives us the err msg we can match and then do custom handling. This does not prevent the printing of the help message however.

```bash
> odo create python --starter
yay i can detect starter was not used
Error: flag needs an argument: --starter
Usage:
  odo create <component_type> [component_name] [flags]

Examples:
  # Create a new Node.JS component with existing sourcecode as well as specifying a name
  odo create nodejs mynodejs
  
  # Name is not required and will be automatically generated if not passed
  odo create nodejs
  
  # List all available components before deploying
  odo catalog list components
  odo create java-quarkus
  
  # Download an example devfile and application before deploying
  odo create nodejs --starter
  
  # Using a specific devfile
  odo create mynodejs --devfile ./devfile.yaml
  odo create mynodejs --devfile https://raw.githubusercontent.com/odo-devfiles/registry/master/devfiles/nodejs/devfile.yaml
  
  # Create new Node.js component named 'frontend' with the source in './frontend' directory
  odo create nodejs frontend --context ./frontend
  
  # Create new Java component with binary named sample.jar in './target' directory; this flag is specific to S2I
  # DEPRECATED: Use the devfile to create a component
  odo create java:8 --s2i --binary target/sample.jar
  
  # Create new Node.js component with source from remote git repository; this flag is specific to S2I
  # DEPRECATED: Use the devfile to create a component
  odo create nodejs --s2i --git https://github.com/openshift/nodejs-ex.git
  
  # Create new Node.js component with custom ports and environment variables
  odo create nodejs --port 8080,8100/tcp,9100/udp --env key=value,key1=value1
  
  # Create a new Node.js component that is a part of 'myapp' app inside the 'myproject' project
  odo create nodejs --app myapp --project myproject

Flags:
      --app string             Application, defaults to active application
  -b, --binary string          Create a binary file component component using given artifact. Works only with Java components. File needs to be in the context directory. Specific to S2I components. (DEPRECATED: support for the S2I components has been deprecated, consider creating a devfile component instead.)
      --context string         Use given context directory as a source for component settings
      --devfile string         Path to the user specified devfile
      --env strings            Environmental variables for the component. For example --env VariableName=Value
  -g, --git string             Create a git component using this repository. Specific to S2I components. (DEPRECATED: support for the S2I components has been deprecated, consider creating a devfile component instead.)
  -h, --help                   Help for create
      --now                    Push changes to the cluster immediately
  -p, --port strings           Ports to be used when the component is created (ex. 8080,8100/tcp,9100/udp)
      --project string         Project, defaults to active project
  -r, --ref string             Use a specific ref e.g. commit, branch or tag of the git repository (only valid for --git components) (DEPRECATED: support for the S2I components has been deprecated, consider creating a devfile component instead.)
      --registry string        Create devfile component from specific registry
      --s2i                    Enforce S2I type components (DEPRECATED: support for the S2I components has been deprecated, consider creating a devfile component instead.)
      --starter string         Download a project specified in the devfile
      --starter-token string   Token to be used when downloading starter project
      --token string           Token to be used when downloading devfile from the devfile path that is specified via --devfile

Additional Flags:
      --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
  -o, --o string             Specify output format, supported format: json
  -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging

 ✗  flag needs an argument: --starter

```

So the error itself is printed as part of help message as well as at the bottom. We can either list the starters before the help(any printf we do in the func comes before the help message sa can be seen above) or as part of wrapped error message. wdyt would be the best approach

**Comment by feloy** [comment]
Did you try to make the handler (`f func(*Command, error) error`) return a nil error, to see if the error message is still displayed?

**Comment by mohammedzee1000** [comment]
Ah returning nil jus helps would would it not still be odd behavior?

**Comment by mohammedzee1000** [comment]
Also would it not be easier to point user to use `odo catalog describe component <name>` to find list com starters. This way we wont have to recreate context in handler

**Comment by feloy** [comment]
With this handler, you can get the list of starters (or the command to get it) after the body of the error:

```
	componentCreateCmd.SetFlagErrorFunc(func(c *cobra.Command, err error) error {
		if "test if --starter error ..." {
				return fmt.Errorf("%w: you can get the list of possible values with the comamnd ...", err)
		}
		return err
	})
```

**Comment by mohammedzee1000** [comment]
That was what I was thinking, thanks
